### PR TITLE
Stored Procedures

### LINKED ISSUE #19: Support for Stored Procedure command type
Current version of MySqlCommand only supports CommandType = Text. Supporting stored procedure commands is a must for any realistic migration from the official MySQL Connector.

ps: nice work so far!

### PR DESCRIPTION
**Pull request by caleblloyd**
- Fixes #19 
- Implements Stored Procedure Cache on a per-connection basis in >= [MySql 5.5.3](http://dev.mysql.com/doc/refman/5.5/en/parameters-table.html)
  - Fixes up Parameter Order, Direction, and Data Types to match the procedure on the Server
- Works without Stored Procedure Cache on < MySql 5.5.3, but caller is responsible for matching c# param Order, Direction, and Data Types to Server params

### CODE REVIEW

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+		{
+			m_database.Connection.Close();
+		}
+
+		[Fact]
+		public async Task StoredProcedureBasic()
+		{
+			using (var cmd = m_database.Connection.CreateCommand())
+			{
+				cmd.CommandText = @"DROP PROCEDURE IF EXISTS simple;
```
Comment (caleblloyd) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Failing to create in `MySql.Data`

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+				Assert.Equal("simple", cmd.Parameters["@result"].Value);
+			}
+		}
+
+		[Fact]
+		public async Task StoredProcedureComplex()
+		{
+			using (var cmd = m_database.Connection.CreateCommand())
+			{
+				cmd.CommandText = @"DROP PROCEDURE IF EXISTS complex;
```
Comment (caleblloyd) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Failing to create in `MySql.Data`

**Comment by caleblloyd** [comment]
I want to separate these methods into an interface:

```csharp
interface ICommandExecutor

async Task<int> ExecuteNonQueryAsync(string commandText, MySqlParameterCollection parameterCollection, IOBehavior ioBehavior, CancellationToken cancellationToken)

async Task<object> ExecuteScalarAsync(string commandText, MySqlParameterCollection parameterCollection, IOBehavior ioBehavior, CancellationToken cancellationToken)

async Task<DbDataReader> ExecuteReaderAsync(string commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken)
```

Then create 2 implementations:

- `TextCommandExecutor`
- `StoredProcedureCommandExecutor`

@bgrainger do you have a recommendation for a namespace I should put these in?

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs [hunk 3]
```diff
@@ -111,7 +114,70 @@ protected override DbParameter CreateDbParameter()
 
 		internal async Task<int> ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
 		{
-			using (var reader = (MySqlDataReader) await ExecuteReaderAsync(CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+			if (CommandType == CommandType.StoredProcedure)
+				return await ExecuteStoredProcedureAsync(ioBehavior, cancellationToken);
+
+			return await ExecuteTextCommandNonQueryAsync(CommandText, m_parameterCollection, ioBehavior, cancellationToken);
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: Missing `.ConfigureAwait(false)` on both lines.

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs [hunk 4]
```diff
@@ -111,7 +114,70 @@ protected override DbParameter CreateDbParameter()
...
+				}
+
+				if (parameter.Direction == ParameterDirection.ReturnValue)
+					returnParameter = parameter;
+				else
+					argParamNames.Add(parameter.ParameterName);
+			}
+
+			int affectedRows;
+			var commandText = "CALL " + CommandText + "(" + string.Join(", ", argParamNames) +")";
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: Does `argParamNames` need escaping/quoting?

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs [hunk 5]
```diff
@@ -111,7 +114,70 @@ protected override DbParameter CreateDbParameter()
...
+			if (outParams.Count > 0)
+			{
+				commandText = "SELECT " + string.Join(", ", outParamNames);
+				using (var reader = (MySqlDataReader) await ExecuteTextCommandReaderAsync(commandText, new MySqlParameterCollection(), CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+				{
+					await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					for (var i = 0; i < outParams.Count; i++)
+					{
+						var parameter = outParams[i];
+						parameter.Value = reader.GetValue(i);
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: Nitpick: unless you're going to use it for something else, I'd probably combine these into one line: `outParams[i].Value = reader.GetValue(i);`.

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs [hunk 6]
```diff
@@ -128,8 +194,16 @@ protected override DbParameter CreateDbParameter()
 
 		internal async Task<object> ExecuteScalarAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
 		{
+			if (CommandType == CommandType.StoredProcedure)
+				throw new MySqlException("Stored Procedures should be run with ExecuteNonQuery or ExecuteNonQueryAsync");
+
+			return await ExecuteTextCommandScalarAsync(CommandText, m_parameterCollection, ioBehavior, cancellationToken);
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: `.ConfigureAwait(false)`

File: tests/SideBySide.New/StoredProcedureTests.cs (hunk 1 above)
Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: FWIW, the exception is being thrown from line 45, i.e., `ExecuteNonQueryAsync` on the stored procedure.

Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: It appears that when you use a `ReturnValue` parameter, Connector/Net assumes you are calling a function, not a stored procedure. The following code (inspired by https://dev.mysql.com/doc/connector-net/en/connector-net-tutorials-stored-procedures.html) works with Connector/Net:

```
			cmd.CommandText = "simple";
			cmd.CommandType = CommandType.StoredProcedure;
			using (var reader = await cmd.ExecuteReaderAsync())
			{
				Assert.True(await reader.ReadAsync());
				Assert.Equal("simple", reader.GetString(0));
			}
```

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+					DbType = DbType.String,
+					Direction = ParameterDirection.Output,
+				});
+				cmd.Parameters.Add(new MySqlParameter
+				{
+					ParameterName = "@result",
+					DbType = DbType.String,
+					Direction = ParameterDirection.ReturnValue,
+				});
+				await cmd.ExecuteNonQueryAsync();
```
Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Again, I think you need to `ExecuteReaderAsync` here and read the (single) row that is returned.

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs (hunk 3 above)
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: Thanks, I'll definitely do a global grep for this before I push the final changeset

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs (hunk 5 above)
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: Oh yea.  I had a foreach loop here at first, but it it wasn't detecting `parameter` as a `MySqlParameter` so I was having to use `as MySqlParameter`.  Changed to a for loop and it worked.  I guess `MySqlParameterCollection` should probably implement `IEnumerable<MySqlParameter>`

**Comment by bgrainger** [comment]
The interfaces could go in `MySql.Data.MySqlClient` (I assume they would be `internal`).

File: src/MySqlConnector/MySqlClient/MySqlCommand.cs (hunk 4 above)
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/MySqlCommand.cs]: I thought about this too.  I think that all OUT parameters need to be [User Defined Vars](http://dev.mysql.com/doc/refman/5.7/en/user-variables.html), which all begin with `@`.  I think the caller should escape them per documentation if they want them escaped:

> User variables are written as @var_name, where the variable name var_name consists of alphanumeric characters, ., _, and $. A user variable name can contain other characters if you quote it as a string or identifier (for example, @'my-var', @"my-var", or @`my-var`).

Also, sub-queries are technically valid:

`CALL complex((SELECT 1), (SELECT 2), (SELECT "name"), @a, @b, @c, @d, @e)`

This would mean somebody would have to set their IN `MySqlParameter.ParameterName = "(SELECT 1)"`.  They should probably be working with Text Commands at this point anyways though.

**Comment by caleblloyd** [comment]
They would be `internal`.  Should I create a sub-namespace, say `MySql.Data.MySqlClient.CommandExecutor` that contains all relevant files.  Right now I'm thinking there will be 4 and am worried about cluttering up the `MySql.Data.MySqlClient` ns.  Also it's hard to tell what's related when the first letter puts them far apart in the alphabetized directory listing with a bunch of other unrelated classes.

- `interface ICommandExecutor`
- `abstract BaseCommandExecutor`
- `TextCommandExecutor extends BaseCommandExecutor`
- `StoredProcedureCommandExecutor extends BaseCommandExecutor`

**Comment by bgrainger** [comment]
`MySql.Data.MySqlClient.CommandExecutors` sounds good. ("[Consider using plural namespace names where appropriate."](https://msdn.microsoft.com/en-us/library/ms229026.aspx))

File: src/MySqlConnector/MySqlClient/CommandExecutors/StoredProcedureCommandExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+				commandText = "SELECT " + string.Join(", ", outParamNames);
+				using (var reader = await DoExecuteReaderAsync(commandText, new MySqlParameterCollection(), CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+				{
+					await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					for (var i = 0; i < outParams.Count; i++)
+						outParams[i].Value = reader.GetValue(i);
+				}
+			}
+
+			return null;
```
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/CommandExecutors/StoredProcedureCommandExecutor.cs]: I think `MySql.Data` may actually return a reader here.  How could we return a reader since it is used up when getting the return value?

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	SELECT 'simple';
+END";
+				await cmd.ExecuteNonQueryAsync();
+			}
+
+			using (var cmd = m_database.Connection.CreateCommand())
+			{
+				cmd.CommandText = "simple";
+				cmd.CommandType = CommandType.StoredProcedure;
+// MySql.Data makes the assumption that Stored Procedures with ParameterDirection.ReturnValue are functions
```
Comment (caleblloyd) [inline tests/SideBySide.New/StoredProcedureTests.cs]: I don't like the assumption `MySql.Data` makes about Stored Procedure vs. Function based off ParameterDirection.ReturnValue

Could we implement an extension of [System.Data.CommandType](https://msdn.microsoft.com/en-us/library/system.data.commandtype(v=vs.110).aspx) and include a `MySqlCommandType.Function` enum?

Comment (caleblloyd) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Actually, since enums can't be extended, we could introduce a `StoredProcedureType` enum with options of `Procedure` and `Function`. It would be exposed in `MySqlCommand.StoredProcedureType`

Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Most stored procedures will return results in their result set (read by `ReadAsync`), even if it's a single row. So MySQL's assumption may not be that strange here. I'd need to do more investigation to be sure it's worth breaking API compatibility in this situation.

**Comment by caleblloyd** [comment]
Latest change set implements stored procedures 1:1 with `MySql.Data`

- Check for a parameter with `ParameterDirection.ReturnValue` and call function, otherwise call stored procedure
- Added a `MySqlCommand.ReaderClosed` function that sets stored procedure `Out` parameter values after a stored procedure's reader is closed.
- Stored procedures returning result sets is now supported

**Comment by bgrainger** [comment]
This code needs more test coverage to verify it's handling all the edge cases correctly (and there are probably a lot). For example, I just wrote some naive code to test stored procedures: 763716410130dbbe15622a84ade7a6e250288506

It succeeds with Connector/Net but fails with this code. (I haven't debugged it yet.)

Additionally, this code seems to be more sensitive to parameter names than Connector/Net; `@factor` works but `factor` doesn't.

**Comment by caleblloyd** [comment]
It looks like `MySql.Data` makes up it's own user variable names:

https://github.com/mysql/mysql-connector-net/blob/6.9/Source/MySql.Data/StoredProcedure.cs#L179

I'll adjust the code to do this, which should eliminate problems stemming from what users choose to name their parameters.

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+				cmd.CommandType = CommandType.StoredProcedure;
+				cmd.Parameters.Add(new MySqlParameter { ParameterName = paramaterName, Value = 3 });
+				using (var reader = await cmd.ExecuteReaderAsync())
+				{
+					Assert.True(await reader.ReadAsync());
+					Assert.Equal("six", reader.GetString(0));
+					Assert.True(await reader.ReadAsync());
+					Assert.Equal("three", reader.GetString(0));
+					Assert.False(await reader.ReadAsync());
+					//Assert.False(await reader.NextResultAsync()); // <-- This passes on MySql.Data but not MySqlConnector, why?
```
Comment (caleblloyd) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Not sure why this assertion would fail on MySqlConnector.  Is there really another empty result set getting sent by the server?

**Comment by caleblloyd** [comment]
Test coverage now includes all parameter directions (in, out, inout, and return)

`MySql.Data` checks parameter type names match function parameter names in the stored procedure.  It has code to retrieve the stored procedure from the database, parse the parameter names, and enforce that the C# parameter names match.

Our current implementation in this PR just passes positional parameters.  It is up to the caller to name them, choose the right parameter direction, choose the right DbType, and add them to the parameter collection in the correct order.

[This function](https://github.com/mysql/mysql-connector-net/blob/6.9/Source/MySql.Data/StoredProcedure.cs#L114) is where `MySql.Data` performs all of these operations.  Do we want to support all of this?

Current State Pros / Cons:
- Pro: We save a round trip to the Database every time a procedure is called (don't have to get procedure parameters)
- Con: We don't align data types for parameters (this doesn't happen for text SQL queries anyways)
- Con: We don't check to see if all parameters are defined with the proper direction
- Con: We aren't 1:1 backward compatible with `MySql.Data`

**Comment by caleblloyd** [comment]
It's probably best to mirror what `MySql.Data` does. They do cache the procedure parameters for the life of the connection. INFORMATION_SCHEMA only started supporting parameter names after MySQL 5.5.3 so I'll add a version check. Anything lower than that will fallback on using positional parameters.

File: src/MySqlConnector/MySqlClient/Caches/NormalizedName.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Text.RegularExpressions;
+
+namespace MySql.Data.MySqlClient.Caches
+{
+	internal class NormalizedName
+	{
+		private static readonly Regex NameRe = new Regex(@"^\s*(?:`([^`]+)`|([^\.]+)?)\s*\.?\s*(?:`([^`]+)`|([^\.]+)?)?\s*$",
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/NormalizedName.cs]: There are valid (but pathological) MySQL identifiers that will break this regex. 

> Identifier quote characters can be included within an identifier if you quote the identifier. If the character to be included within the identifier is the same as that used to quote the identifier itself, then you need to double the character.

http://dev.mysql.com/doc/refman/5.7/en/identifiers.html

May not be worth fixing, since it's such an edge case but FYI (in case you want to make a note or a known issue about it).

File: src/MySqlConnector/MySqlClient/Caches/Procedure.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+			{
+				ParameterName = "@component",
+				DbType = DbType.String,
+				Value = normalized.Component
+			});
+
+			var parameters = new List<Parameter>();
+			using (var reader = (MySqlDataReader) await cmd.ExecuteReaderAsync(CommandBehavior.Default, ioBehavior, cancellationToken))
+			{
+				while (reader.Read())
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/Procedure.cs]: Do you want `ReadAsync` here?

File: src/MySqlConnector/MySqlClient/Caches/Procedure.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+			});
+			cmd.Parameters.Add(new MySqlParameter
+			{
+				ParameterName = "@component",
+				DbType = DbType.String,
+				Value = normalized.Component
+			});
+
+			var parameters = new List<Parameter>();
+			using (var reader = (MySqlDataReader) await cmd.ExecuteReaderAsync(CommandBehavior.Default, ioBehavior, cancellationToken))
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/Procedure.cs]: `.ConfigureAwait(false)` :smile:

File: src/MySqlConnector/MySqlClient/Caches/NormalizedName.cs (hunk 11 above)
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/Caches/NormalizedName.cs]: Wow. I feel sorry for anyone working with a database that contains back ticks in the schema name

File: src/MySqlConnector/MySqlClient/Caches/Procedure.cs (hunk 12 above)
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/Caches/Procedure.cs]: Yes, thanks for catching

File: src/MySqlConnector/MySqlClient/MySqlConnection.cs [hunk 14]
```diff
@@ -198,6 +200,34 @@ internal MySqlSession Session
...
+			{
+				cachedProcedure = await CachedProcedure.FillAsync(ioBehavior, this, normalized.Schema, normalized.Component, cancellationToken).ConfigureAwait(false);
+				m_cachedProcedures[normalized.CacheKey] = cachedProcedure;
+			}
+			catch (MySqlException ex)
+			{
+				if (ex.Number != 1109)
+					throw;
+				// 1109 means information_schema.parameters table not found on MySql < 5.5.3, see http://dev.mysql.com/doc/refman/5.5/en/parameters-table.html
+				m_cachedProceduresSupported = false;
```
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/MySqlConnection.cs]: Opting for an Exception Error Number Check here instead of checking the Server version before the query.  This way we don't have to maintain a list of server types and versions that support `information_schema.parameters` (MySql >= 5.5.3, MariaDB >= whatever, etc).  Is this acceptable?

Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/MySqlConnection.cs]: Apparently their version numbers are supposed to be [1:1 compatible](https://mariadb.com/kb/en/mariadb/mariadb-vs-mysql-compatibility/), so I'll change this to a version check

File: src/MySqlConnector/MySqlClient/Caches/CachedParameter.cs [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+					case "INOUT":
+						Direction = ParameterDirection.InputOutput;
+						break;
+					case "OUT":
+						Direction = ParameterDirection.Output;
+						break;
+				}
+			}
+			Name = name;
+			switch (dataType.ToLower())
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/CachedParameter.cs]: Future enhancement: extract this (and the inverse code in `MySqlDataReader.GetDataTypeName`) to a shared location?

Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/CachedParameter.cs]: Wondering if we need to implement our own `MySqlType` (or similar) that's the superset of MySQL Server data type names, .NET types, `DbType` enumeration values, `ColumnType` enumeration values, etc., etc.

File: src/MySqlConnector/MySqlClient/Types/TypeMapper.cs [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+using System.Linq;
+using MySql.Data.Serialization;
+
+namespace MySql.Data.MySqlClient.Types
+{
+	internal class TypeMapper
+	{
+		internal static TypeMapper Mapper = new TypeMapper();
+
+		private TypeMapper()
```
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/Types/TypeMapper.cs]: Here are the beginnings of a Type Mapper.  I think that we can evolve this namespace (in further commits) to take care of most CLR Type / Column Type / DbType conversions.

File: src/MySqlConnector/MySqlClient/Caches/CachedParameter.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+		public CachedParameter(int ordinalPosition, string mode, string name, string dataType, bool unsigned)
+		{
+			Position = ordinalPosition;
+			if (Position == 0)
+			{
+				Direction = ParameterDirection.ReturnValue;
+			}
+			else
+			{
+				switch (mode.ToLower())
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/CachedParameter.cs]: Need to use `ToLowerInvariant` to avoid the "[Turkish I problem](http://haacked.com/archive/2012/07/05/turkish-i-problem-and-why-you-should-care.aspx/)".

File: src/MySqlConnector/MySqlClient/Caches/CachedProcedure.cs [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+			using (var reader = (MySqlDataReader) await cmd.ExecuteReaderAsync(CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+			{
+				while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
+				{
+					parameters.Add(new CachedParameter(
+						reader.GetInt32(0),
+						!reader.IsDBNull(1) ? reader.GetString(1) : null,
+						!reader.IsDBNull(2) ? reader.GetString(2) : null,
+						reader.GetString(3),
+						reader.GetString(4).ToLower().Contains("unsigned")
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Caches/CachedProcedure.cs]: `reader.GetString(4).IndexOf("unsigned", StringComparison.OrdinalIgnoreCase) != -1`

Avoids culture-sensitive comparison ("Turkish I") and allocating a new string just for substring comparison.

File: src/MySqlConnector/MySqlClient/Types/TypeMapper.cs [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+		{
+			ColumnTypeMapping columnTypeMapping;
+			if (!m_columnTypeMappingLookup.TryGetValue(ColumnTypeMapping.CreateLookupKey(columnTypeName, unsigned, length), out columnTypeMapping) && length != 0)
+				m_columnTypeMappingLookup.TryGetValue(ColumnTypeMapping.CreateLookupKey(columnTypeName, unsigned, 0), out columnTypeMapping);
+			return columnTypeMapping;
+		}
+
+		private Dictionary<Type, DbTypeMapping> m_dbTypeMappingsByClrType = new Dictionary<Type, DbTypeMapping>();
+		private Dictionary<DbType, DbTypeMapping> m_dbTypeMappingsByDbType = new Dictionary<DbType, DbTypeMapping>();
+		private Dictionary<string, ColumnTypeMapping> m_columnTypeMappingLookup = new Dictionary<string, ColumnTypeMapping>();
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/Types/TypeMapper.cs]: Could use `StringComparer.OrdinalIgnoreCase` as the key comparer (to match case-insensitively without having to lowercase all input strings).

Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/Types/TypeMapper.cs]: Ah I see, good tip.  Since everything runs through CreateLookupKey and this is private, I'll just call `ToLowerInvariant()` there (no need to double-tap)

**Comment by caleblloyd** [comment]
I think this is about ready to merge.  @bgrainger do you see any features or test coverage that should be added?  I am going to squash the commits.

**Comment by bgrainger** [comment]
I think I should have time to go through this in a bit more detail tonight.

File: src/MySqlConnector/MySqlClient/MySqlParameter.cs [hunk 20]
```diff
@@ -153,9 +173,16 @@ internal void AppendSqlString(BinaryWriter writer, StatementPreparerOptions opti
 
 		internal static string NormalizeParameterName(string name)
 		{
-			return string.IsNullOrEmpty(name) ? name :
-				name[0] == '?' || name[0] == '@' ? name.Substring(1) :
-				name;
+			name = name.Trim();
+
+			if ((name.StartsWith("@`") || name.StartsWith("?`")) && name.EndsWith("`"))
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlParameter.cs]: Use the overloads that take `StringComparison.Ordinal`. The default behaviour is to be culture-sensitive which is horribly inefficient in CoreCLR on Linux: dotnet/coreclr#5612 dotnet/corefx#5027

File: src/MySqlConnector/MySqlClient/ServerVersions.cs [hunk 21]
```diff
@@ -4,6 +4,10 @@ namespace MySql.Data.MySqlClient
 {
 	internal static class ServerVersions
 	{
+		// https://dev.mysql.com/doc/refman/5.7/en/mysql-reset-connection.html
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/ServerVersions.cs]: 👍 for documentation

File: tests/MySqlConnector.Tests/TypeMapperTests.cs [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+		[InlineData("multilinestring", false, 0, DbType.Binary)]
+		[InlineData("multipolygon", false, 0, DbType.Binary)]
+		[InlineData("datetime", false, 0, DbType.DateTime)]
+		[InlineData("date", false, 0, DbType.DateTime)] // todo: this should be DbType.Date
+		[InlineData("time", false, 0, DbType.Time)]
+		[InlineData("timestamp", false, 0, DbType.DateTimeOffset)]
+		[InlineData("year", false, 0, DbType.Int32)]
+		public void ColumnTypeMappingTest(string columnTypeName, bool unsigned, int length, DbType dbType)
+		{
+			Assert.Equal(dbType, TypeMapper.Mapper.GetDbTypeMapping(columnTypeName, unsigned, length).DbTypes.FirstOrDefault());
```
Comment (bgrainger) [inline tests/MySqlConnector.Tests/TypeMapperTests.cs]: Suggestion: add UPPERCASE test values, or a second test with `columnTypeName.ToUpperInvariant()`?

File: tests/SideBySide.New/StoredProcedureTests.cs [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+			using (var cmd = m_database.Connection.CreateCommand())
+			{
+				cmd.CommandText = "circle";
+				cmd.CommandType = CommandType.StoredProcedure;
+				cmd.Parameters.Add(new MySqlParameter
+				{
+					ParameterName = "@radius",
+					DbType = DbType.Double,
+					Direction = ParameterDirection.Input,
+					Value = (double)1,
```
Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: Suggestion: `Value = 1.0`

File: tests/MySqlConnector.Tests/TypeMapperTests.cs (hunk 22 above)
Comment (caleblloyd) [inline tests/MySqlConnector.Tests/TypeMapperTests.cs]: second assertion with ToUpperInvariant() added

File: src/MySqlConnector/MySqlClient/Caches/CachedProcedure.cs [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+		}
+
+		internal MySqlParameterCollection AlignParamsWithDb(MySqlParameterCollection parameterCollection)
+		{
+			var alignedParams = new MySqlParameterCollection();
+
+			MySqlParameter returnParam = null;
+			for (var i = 0; i < parameterCollection.Count; i++)
+				if (parameterCollection[i].Direction == ParameterDirection.ReturnValue)
+					returnParam = parameterCollection[i];
```
Comment (caleblloyd) [inline src/MySqlConnector/MySqlClient/Caches/CachedProcedure.cs]: Done.  Had to make `MySqlParameterCollection` implement `IEnumerable<MySqlParameter>`

**Comment by caleblloyd** [comment]
Style fixes are in

File: tests/SideBySide.New/StoredProcedureTests.cs (hunk 10 above)
Comment (bgrainger) [inline tests/SideBySide.New/StoredProcedureTests.cs]: There is. https://dev.mysql.com/doc/internals/en/multi-resultset.html says:

> ... and a closing empty resultset, an OK_Packet. ... The trailing OK_Packet is the response to the CALL statement and contains the affected-rows count of the last statement. 

I think the way to address this might be to add extra logic in `NextResultAsync` to detect this empty result set condition (when `CommandType == StoredProcedure`) and return an accurate `false` value.

I can think of one edge case where this will be a problem: someone using Dapper's `QueryMultiple` to call a sproc with multiple result sets. Their code will only read _n_ result sets but Dapper will think there are _n+1_, so will `Cancel` the command because it thinks there was an incomplete read. This will currently throw an exception in this library. (Dapper's regular `Query` for a sproc will work just fine.)

It seems fine to me to ship with this as a known issue and address it in a future version. (Shipping support for stored procedures may turn up other problems too, and we can fix them all at once.)
